From cfb8627a80a334f200f68c2c8f3e384313ebbaf5 Mon Sep 17 00:00:00 2001
From: Raymond Cole <rc@wolog.xyz>
Date: Mon, 28 Oct 2024 00:34:55 +0000
Subject: [PATCH] Avoid unsigned integer underflow in drw_text()

---
 drw.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drw.c b/drw.c
index 344de61..c41e6af 100644
--- a/drw.c
+++ b/drw.c
@@ -248,6 +248,8 @@ drw_text(Drw *drw, int x, int y, unsigned int w, unsigned int h, unsigned int lp
 	} else {
 		XSetForeground(drw->dpy, drw->gc, drw->scheme[invert ? ColFg : ColBg].pixel);
 		XFillRectangle(drw->dpy, drw->drawable, drw->gc, x, y, w, h);
+		if (w < lpad)
+			return x + w;
 		d = XftDrawCreate(drw->dpy, drw->drawable,
 		                  DefaultVisual(drw->dpy, drw->screen),
 		                  DefaultColormap(drw->dpy, drw->screen));
-- 
2.51.0

